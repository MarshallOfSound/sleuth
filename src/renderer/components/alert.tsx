import { observer } from 'mobx-react';
import { SleuthState } from '../state/sleuth';
import * as React from 'react';

export interface AlertProps {
  text: string;
  level?: 'error' | 'generic' | 'warning' | 'success' | 'info';
  state?: SleuthState;
}

export interface AlertState {}

@observer
export class Alert extends React.Component<AlertProps, AlertState> {
  constructor(props: AlertProps) {
    super(props);

    this.onclick = this.onclick.bind(this);
  }

  public getIcon() {
    const { level } = this.props;
    let icon = 'ts_icon_slack_pillow';

    if (level === 'warning' || level === 'error') {
      icon = 'ts_icon_warning';
    } else if (level === 'success') {
      icon = 'ts_icon_check_circle_o';
    } else if (level === 'info') {
      icon = 'ts_icon_info_circle';
    }

    return <i className={`ts_icon ${icon}`} />;
  }

  public onclick() {
    if (this.props.state) {
      this.props.state.webAppLogsWarningDismissed = true;
    }
  }

  public render(): JSX.Element | null {
    if (this.props.state && this.props.state.webAppLogsWarningDismissed) return null;

    const { level, text } = this.props;
    const levelClassName = level === 'generic' ? '' : level;
    const icon = this.getIcon();

    return (
      <div onClick={this.onclick} className={`alert alert_${levelClassName}`}>
        {icon}
        {text}
      </div>
    );
  }
}
