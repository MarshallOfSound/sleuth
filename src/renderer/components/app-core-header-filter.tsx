import { observer } from 'mobx-react';
import React from 'react';
import debounce from 'debounce';
import {
  Button,
  Classes,
  NavbarDivider,
  NavbarGroup,
  ButtonGroup,
  InputGroup,
  Popover,
  Menu,
  Position
} from '@blueprintjs/core';
import { DateRangeInput } from '@blueprintjs/datetime';

import { SleuthState } from '../state/sleuth';

export interface FilterProps {
  state: SleuthState;
}

export interface FilterState {
}

@observer
export class Filter extends React.Component<FilterProps, Partial<FilterState>> {
  constructor(props: FilterProps) {
    super(props);

    this.onFilterToggle = this.onFilterToggle.bind(this);
    this.toggleSearchResultVisibility = this.toggleSearchResultVisibility.bind(this);
    this.onSearchChange = debounce(this.onSearchChange.bind(this), 700);
    this.onDateRangeChange = this.onDateRangeChange.bind(this);
    this.renderFilter = this.renderFilter.bind(this);
  }

  public onSearchChange(value: string) {
    this.props.state.search = value;
  }

  public onFilterToggle(level: string) {
    if (this.props.state.levelFilter![level] !== undefined) {
      const filter = {...this.props.state.levelFilter};
      filter[level] = !filter[level];

      this.props.state.levelFilter = filter;
    }
  }

  public onSearchIndexChange(change: number) {
    this.props.state.searchIndex = this.props.state.searchIndex + change;
  }

  public onDateRangeChange(range: [ Date?, Date? ]) {
    this.props.state.dateRange = {
      from: range[0],
      to: range[1]
    };
  }

  public toggleSearchResultVisibility() {
    this.props.state.showOnlySearchResults = !this.props.state.showOnlySearchResults;
  }

  public renderFilter() {
    const { error, warn, info, debug } = this.props.state.levelFilter!;

    const menu = (
      <Menu>
        <Menu.Item
          active={warn}
          onClick={() => this.onFilterToggle('warn')}
          icon='warning-sign'
          shouldDismissPopover={false}
          text='Warning'
        />
        <Menu.Item
          active={info}
          onClick={() => this.onFilterToggle('info')}
          icon='info-sign'
          shouldDismissPopover={false}
          text='Info'
        />
        <Menu.Item
          active={error}
          onClick={() => this.onFilterToggle('error')}
          icon='error'
          shouldDismissPopover={false}
          text='Error'
        />
        <Menu.Item
          active={debug}
          onClick={() => this.onFilterToggle('debug')}
          icon='code'
          shouldDismissPopover={false}
          text='Debug'
        />
      </Menu>
    );

    return (
      <Popover content={menu} position={Position.BOTTOM}>
        <Button icon='filter-list' text='Filter'/>
      </Popover>
    );
  }

  public render() {
    const { showOnlySearchResults, dateRange } = this.props.state;

    const showOnlySearchResultsButton = (
      <Button
        active={showOnlySearchResults}
        onClick={this.toggleSearchResultVisibility}
        className={Classes.MINIMAL}
        icon={showOnlySearchResults ? 'eye-off' : 'eye-open'}
      />
    );

    return (
      <>
        <NavbarGroup className='FilterGroup'>
          {this.renderFilter()}
        </NavbarGroup>
        <NavbarGroup className='SearchGroup'>
          <NavbarDivider />
          <DateRangeInput
            formatDate={(date) => date.toLocaleString()}
            onChange={this.onDateRangeChange}
            parseDate={(str) => new Date(str)}
            value={[ dateRange.from, dateRange.to ]}
          />
          <Button
            icon='cross'
            style={{ marginLeft: '5px' }}
            onClick={() => this.onDateRangeChange([])}
          />
          <NavbarDivider />
          <InputGroup
            leftIcon='search'
            placeholder='Search'
            rightElement={showOnlySearchResultsButton}
            onChange={(e: React.FormEvent) => this.onSearchChange((e.target as any).value)}
          />
          <NavbarDivider />
          <ButtonGroup>
            <Button icon='arrow-left' onClick={() => this.onSearchIndexChange(-1)} />
            <Button icon='arrow-right' onClick={() => this.onSearchIndexChange(1)} />
          </ButtonGroup>
        </NavbarGroup>
      </>
    );
  }
}
