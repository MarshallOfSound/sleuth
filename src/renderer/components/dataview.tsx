import * as React from 'react';
import * as classNames from 'classnames';
import * as dirtyJSON from 'jsonic';
import JSONTree from 'react-json-tree';
import * as moment from 'moment';

import { LogEntry } from '../interfaces';

export interface DataViewProps {
  isVisible: boolean;
  entry?: LogEntry;
  height?: number;
  toggle: Function;
  logEntry?: LogEntry;
}

export interface DataViewState {

}

export class DataView extends React.Component<DataViewProps, DataViewState> {
  constructor(props: DataViewProps) {
    super(props);

    this.onDoubleClick = this.onDoubleClick.bind(this);
  }

  /**
   * Toggle the whole data view.
   */
  public onDoubleClick() {
    this.props.toggle();
  }

  /**
   * Takes a meta string (probably dirty JSON) and attempts to pretty-print it.
   *
   * @param {string} meta
   * @returns {(JSX.Element | null)}
   */
  public renderMeta(meta: string): JSX.Element | null {
    if (!meta) {
      return null;
    }

    try {
      const data = dirtyJSON(meta);

      if (data) {
        return <JSONTree data={data} theme={this.getTheme()} />;
      } else {
        return <code>{meta}</code>;
      }
    } catch (e) {
      return <code>{meta}</code>;
    }
  }

  /**
   * Renders a single log entry, ensuring that people can scroll around and still now what log entry they're looking at.
   *
   * @param {LogEntry} logEntry
   * @returns {(JSX.Element | null)}
   */
  public renderLogEntry(logEntry: LogEntry): JSX.Element | null {
    const { level, logType, message, timestamp } = logEntry;
    const type = `${logType.charAt(0).toUpperCase() + logType.slice(1)} Process`;
    const datetime = logEntry.moment ? moment(logEntry.moment).format("dddd, MMMM Do YYYY, h:mm:ss a") : timestamp;

    return (
      <div className='DataView-LogEntry'>
        <div className='DataView-MetaInfo'>
          <div className='DataView-Moment'>{datetime}</div>
          <div className='DataView-LogType'>
            	<i className="ts_icon ts_icon_volume_medium"></i>{level}
              <i className="ts_icon ts_icon_volume_medium"></i>{type}
          </div>
        </div>
        <div className='DataView-Message'>{message}</div>
      </div>
    )
  }

  public render(): JSX.Element | null {
    const { entry, isVisible, height, logEntry } = this.props;
    const style = { height: `${height || 300}px` };
    const meta = entry ? entry.meta : '';
    const className = classNames('DataView', { IsVisible: isVisible });
    const logEntryInfo = logEntry ? this.renderLogEntry(logEntry) : null;
    const prettyMeta = this.renderMeta(meta);

    return (
      <div className={className} style={style} onDoubleClick={this.onDoubleClick}>
        {logEntryInfo}
        {prettyMeta}
      </div>
    );
  }

  /**
   * A super-cool Base16 theme using Slack's 2016 colors.
   *
   * @returns {Object}
   */
  public getTheme() {
    return {
      base00: '#2C2D30',
      base01: '#555459',
      base02: '#8B898F',
      base03: '#88919B',
      base04: '#9e9ea6',
      base05: '#FBFBFA',
      base06: '#F9F9F9',
      base07: '#ffffff',
      base08: '#e32072',
      base09: '#F96A38',
      base0A: '#FFA940',
      base0B: '#257337',
      base0C: '#3971ED',
      base0D: '#3971ED',
      base0E: '#71105F',
      base0F: '#4d6dc3'
    };
  }
}
