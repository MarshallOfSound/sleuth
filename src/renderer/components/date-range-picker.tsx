import { observer } from 'mobx-react';
import * as React from 'react';
import DayPicker from 'react-day-picker';
import * as classNames from 'classnames';

import { SleuthState } from '../state/sleuth';

export interface DateRangePickerProps {
  state: SleuthState;
}

export interface DateRangePickerState {
  isOpen: boolean;
}

@observer
export class DateRangePicker extends React.Component<DateRangePickerProps, Partial<DateRangePickerState>> {
  constructor(props: DateRangePickerProps) {
    super(props);

    this.state = {
      isOpen: false
    };

    this.onToggleOpen = this.onToggleOpen.bind(this);
    this.onDayClick = this.onDayClick.bind(this);
  }

  public onToggleOpen() {
    this.setState({ isOpen: !this.state.isOpen });
  }

  public onDayClick(day: Date) {
    const range = DayPicker.DateUtils.addDayToRange(day, {
      from: this.props.state.dateFrom!,
      to: this.props.state.dateTo!
    });

    this.props.state.dateFrom = range.from;
    this.props.state.dateTo = range.to;
  }

  public render() {
    const { isOpen } = this.state;
    const { dateFrom, dateTo } = this.props.state;
    const isEngaged = dateFrom && dateTo;
    const picker = isOpen ? this.renderPicker() : null;

    return (
      <>
        <a
          className={classNames({ Engaged: isEngaged })}
          onClick={() => this.onToggleOpen()}
        >
          <i className='ts_icon ts_icon_calendar' />
          <span className='block label'>Select Dates</span>
        </a>
        {picker}
      </>
    );
  }

  public renderPicker() {
    const { dateFrom, dateTo } = this.props.state;
    const modifiers = { start: dateFrom, end: dateTo };
    const selectedDays = [ dateFrom, { from: dateFrom, to: dateTo } ];

    return (
      <div className='DateRangePicker'>
        <DayPicker
          className='Selectable'
          numberOfMonths={2}
          selectedDays={selectedDays as any}
          modifiers={modifiers}
          onDayClick={this.onDayClick}
        />
      </div>
    );
  }
}
