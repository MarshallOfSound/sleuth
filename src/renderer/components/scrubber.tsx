import React from 'react';

export interface ScrubberProps {
  onResizeHandler: (newHeight: number) => void;
  elementSelector: string;
}

export interface ScrubberState {
  startY: number;
  startHeight: number;
}

export class Scrubber extends React.Component<ScrubberProps, ScrubberState> {
  constructor(props: ScrubberProps) {
    super(props);

    this.mouseDownHandler = this.mouseDownHandler.bind(this);
    this.mouseMoveHandler = this.mouseMoveHandler.bind(this);
    this.mouseUpHandler = this.mouseUpHandler.bind(this);
  }

  public mouseMoveHandler(e: any) {
    const { startHeight, startY } = this.state;
    const newHeight = startHeight + e.clientY - startY;
    this.props.onResizeHandler(newHeight);
  }

  public mouseDownHandler(e: any) {
    const resizeTarget = document.getElementById(this.props.elementSelector);

    if (!resizeTarget) return;

    this.setState({
      startY: e.clientY,
      startHeight: parseInt(document.defaultView!.getComputedStyle(resizeTarget).height!, 10)
    });

    document.documentElement!.addEventListener('mousemove', this.mouseMoveHandler, false);
    document.documentElement!.addEventListener('mouseup', this.mouseUpHandler, false);
  }

  public mouseUpHandler() {
    document.documentElement!.removeEventListener('mousemove', this.mouseMoveHandler, false);
    document.documentElement!.removeEventListener('mouseup', this.mouseUpHandler, false);
  }

  public render() {
    return (
      <div className='Scrubber' onMouseDown={this.mouseDownHandler} />
    );
  }
}
