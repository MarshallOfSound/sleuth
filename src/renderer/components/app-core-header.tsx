import { observer } from 'mobx-react';
import { SleuthState } from '../state/sleuth';
import React from 'react';
import { Alignment, Button, ButtonGroup, Navbar, NavbarGroup } from '@blueprintjs/core';

import { Filter } from './app-core-header-filter';

export interface AppCoreHeaderProps {
  state: SleuthState;
}

export interface AppCoreHeaderState {}

@observer
export class AppCoreHeader extends React.Component<AppCoreHeaderProps, AppCoreHeaderState> {
  public render() {
    const {
      isSidebarOpen,
      isDarkMode,
      isSpotlightOpen,
      toggleDarkMode,
      toggleSidebar,
      toggleSpotlight,
      reset
    } = this.props.state;
    const sidebarIcon = isSidebarOpen ? 'menu-closed' : 'menu-open';

    return (
      <Navbar className='AppHeader'>
        <NavbarGroup align={Alignment.LEFT}>
          <ButtonGroup>
            <Button onClick={() => reset(true)} icon='home' />
            <Button active={!isSidebarOpen} onClick={toggleSidebar} icon={sidebarIcon} />
            <Button active={isDarkMode} onClick={toggleDarkMode} icon='moon' />
            <Button active={isSpotlightOpen} onClick={toggleSpotlight} icon='geosearch' />
          </ButtonGroup>
        </NavbarGroup>
        <Filter state={this.props.state} />
      </Navbar>
    );
  }
}
