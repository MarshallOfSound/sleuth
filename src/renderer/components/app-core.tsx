import { getFirstLogFile } from '../../utils/get-first-logfile';
import { isMergedLogFile, isProcessedLogFile, isUnzippedFile } from '../../utils/is-logfile';
import { observer } from 'mobx-react';
import { sleuthState, SleuthState } from '../state/sleuth';
import * as React from 'react';
import * as classNames from 'classnames';

import { ipcRenderer, remote } from 'electron';
import { UnzippedFile, UnzippedFiles } from '../unzip';
import { getTypesForFiles, mergeLogFiles, processLogFiles } from '../processor';
import {
  LevelFilter,
  MergedFilesLoadStatus,
  MergedLogFile,
  MergedLogFiles,
  ProcessedLogFile,
  ProcessedLogFiles
} from '../interfaces';
import { AppCoreHeader } from './app-core-header';
import { Sidebar } from './sidebar';
import { Loading } from './loading';
import { LogContent } from './log-content';

const debug = require('debug')('sleuth:appCore');

export interface CoreAppProps {
  state: SleuthState;
  unzippedFiles: UnzippedFiles;
}

export interface CoreAppState {
  sidebarIsOpen: boolean;
  processedLogFiles: ProcessedLogFiles;
  selectedLogFile?: ProcessedLogFile | MergedLogFile | UnzippedFile;
  mergedLogFiles?: MergedLogFiles;
  loadingMessage: string;
  loadedLogFiles: boolean;
  loadedMergeFiles: boolean;
  filter: LevelFilter;
  search?: string;
}

@observer
export class CoreApplication extends React.Component<CoreAppProps, Partial<CoreAppState>> {
  constructor(props: CoreAppProps) {
    super(props);

    this.state = {
      sidebarIsOpen: true,
      processedLogFiles: {
        browser: [],
        renderer: [],
        webview: [],
        webapp: [],
        state: [],
        call: []
      },
      loadingMessage: '',
      loadedLogFiles: false,
      loadedMergeFiles: false
    };

    this.toggleSidebar = this.toggleSidebar.bind(this);
    this.selectLogFile = this.selectLogFile.bind(this);

    ipcRenderer.on('processing-status', (_event: any, loadingMessage: string) => {
      this.setState({ loadingMessage });
    });
  }

  /**
   * Once the component has mounted, we'll start processing files.
   */
  public componentDidMount() {
    this.processFiles();
  }

  public componentWillUnmount() {
    ipcRenderer.removeAllListeners('processing-status');
  }

  /**
   * Take an array of processed files (for logs) or unzipped files (for state files)
   * and add them to the state of this component.
   *
   * @param {(Array<ProcessedLogFile|UnzippedFile>)} files
   * @param {string} logType
   */
  public addFilesToState(files: Array<ProcessedLogFile|UnzippedFile>, logType: string) {
    const { processedLogFiles } = this.state;
    const newProcessedLogFiles: ProcessedLogFiles = { ...processedLogFiles as ProcessedLogFiles };
    newProcessedLogFiles[logType] = newProcessedLogFiles[logType].concat(files);

    this.setState({
      processedLogFiles: newProcessedLogFiles
    });
  }

  /**
   * Process files - most of the work happens over in ../processor.ts.
   */
  public async processFiles() {
    const { unzippedFiles } = this.props;

    const sortedUnzippedFiles = getTypesForFiles(unzippedFiles);

    if (Object.keys(sortedUnzippedFiles).map((k) => sortedUnzippedFiles[k]).every((s) => s.length === 0)) {
      const title = 'Huh, weird logs!';
      const message = 'Sorry, Sleuth does not understand the files. It seems like there are no Slack logs here.';

      remote.dialog.showMessageBox({ type: 'error', title, message }, () => {
        remote.getCurrentWindow().reload();
      });
    }

    this.addFilesToState(sortedUnzippedFiles.state, 'state');

    await processLogFiles(sortedUnzippedFiles.renderer)
      .then((newFiles: Array<ProcessedLogFile>) => this.addFilesToState(newFiles, 'renderer'));
    await processLogFiles(sortedUnzippedFiles.browser)
      .then((newFiles: Array<ProcessedLogFile>) => this.addFilesToState(newFiles, 'browser'));
    await processLogFiles(sortedUnzippedFiles.webapp)
      .then((newFiles: Array<ProcessedLogFile>) => this.addFilesToState(newFiles, 'webapp'));
    await processLogFiles(sortedUnzippedFiles.webview)
      .then((newFiles: Array<ProcessedLogFile>) => this.addFilesToState(newFiles, 'webview'));
    await processLogFiles(sortedUnzippedFiles.call)
      .then((newFiles: Array<ProcessedLogFile>) => this.addFilesToState(newFiles, 'call'));

    const { selectedLogFile, processedLogFiles } = this.state;
    if (!selectedLogFile && processedLogFiles) {
      this.props.state.selectedLogFile = getFirstLogFile(processedLogFiles);
      this.setState({ loadedLogFiles: true });
    } else {
      this.setState({ loadedLogFiles: true });
    }

    // We're done processing the files, so let's get started on the merge files.
    this.processMergeFiles();
  }

  /**
   * Update this component's status with a merged logfile
   *
   * @param {MergedLogFile} mergedFile
   */
  public setMergedFile(mergedFile: MergedLogFile) {
    const { mergedLogFiles } = this.state;
    const newMergedLogFiles = { ...mergedLogFiles as MergedLogFiles };

    debug(`Merged log file for ${mergedFile.logType} now created!`);
    newMergedLogFiles[mergedFile.logType] = mergedFile;
    this.setState({ mergedLogFiles: newMergedLogFiles });
  }

  /**
   * Kick off merging of all the log files
   */
  public async processMergeFiles() {
    const { processedLogFiles } = this.state;

    if (processedLogFiles) {
      await mergeLogFiles(processedLogFiles.browser, 'browser').then((r) => this.setMergedFile(r));
      await mergeLogFiles(processedLogFiles.renderer, 'renderer').then((r) => this.setMergedFile(r));
      await mergeLogFiles(processedLogFiles.webview, 'webview').then((r) => this.setMergedFile(r));
      await mergeLogFiles(processedLogFiles.call, 'call').then((r) => this.setMergedFile(r));

      const merged = this.state.mergedLogFiles as MergedLogFiles;

      mergeLogFiles([merged.browser, merged.renderer, merged.webview, merged.call], 'all').then((r) => this.setMergedFile(r));
    }
  }

  /**
   * Toggle the sidebar.
   */
  public toggleSidebar() {
    this.setState({ sidebarIsOpen: !this.state.sidebarIsOpen });
  }

  /**
   * Select a log file. This is a more complex operation than one might think -
   * mostly because we might need to create a merged file on-the-fly.
   *
   * @param {ProcessedLogFile} logFile
   * @param {string} [logType]
   */
  public selectLogFile(logFile: ProcessedLogFile | UnzippedFile | null, logType?: string): void {
    if (!logFile && logType) {
      const { mergedLogFiles } = this.state;

      debug(`Selecting log type ${logType}`);

      if (mergedLogFiles && mergedLogFiles[logType]) {
        this.props.state.selectedLogFile = mergedLogFiles[logType];
      }
    } else if (logFile) {
      const name = isProcessedLogFile(logFile) ? logFile.logType : logFile.fileName;
      debug(`Selecting log file ${name}`);

      this.props.state.selectedLogFile = logFile;
    }
  }

  /**
   * Return the file name of the currently selected file.
   *
   * @returns {string}
   */
  public getSelectedFileName(): string {
    const { selectedLogFile } = this.props.state;

    if (isProcessedLogFile(selectedLogFile)) {
      return selectedLogFile.logFile.fileName;
    } else if (isMergedLogFile(selectedLogFile)) {
      return selectedLogFile.logType;
    } else if (isUnzippedFile(selectedLogFile)) {
      return selectedLogFile.fileName;
    } else {
      return '';
    }
  }

  /**
   * Returns a rounded percentage number for our init process.
   *
   * @returns {number} Percentage loaded
   */
  public getPercentageLoaded(): number {
    const { unzippedFiles } = this.props;
    const processedLogFiles = this.state.processedLogFiles || {};
    const alreadyLoaded = Object.keys(processedLogFiles)
      .map((k) => processedLogFiles[k])
      .reduce((p, c) => p + c.length, 0);
    const toLoad = unzippedFiles.length;

    return Math.round(alreadyLoaded / toLoad * 100);
  }

  /**
   * Real quick: Are we loaded yet?
   *
   * @returns
   */
  public getMergedFilesStatus(): MergedFilesLoadStatus {
    const { mergedLogFiles } = this.state;

    return {
      all: !!(mergedLogFiles && mergedLogFiles.all && mergedLogFiles.all.logEntries),
      browser: !!(mergedLogFiles && mergedLogFiles.browser && mergedLogFiles.browser.logEntries),
      renderer: !!(mergedLogFiles && mergedLogFiles.renderer && mergedLogFiles.renderer.logEntries),
      webview: !!(mergedLogFiles && mergedLogFiles.webview && mergedLogFiles.webview.logEntries),
      webapp: !!(mergedLogFiles && mergedLogFiles.webapp && mergedLogFiles.webapp.logEntries),
      call: !!(mergedLogFiles && mergedLogFiles.call && mergedLogFiles.call.logEntries),
    };
  }

  public render() {
    const { selectedLogFile } = this.props.state;
    const { sidebarIsOpen, processedLogFiles, loadingMessage } = this.state;
    const appCoreClassName = classNames('AppCore');
    const logContentClassName = classNames({ SidebarIsOpen: sidebarIsOpen });
    const selectedLogFileName = this.getSelectedFileName();
    const percentageLoaded = this.getPercentageLoaded();
    const loading = <Loading percentage={percentageLoaded} message={loadingMessage} />;
    const tableOrLoading = selectedLogFile ? <LogContent state={sleuthState} /> : loading;
    const mergedFilesStatus = this.getMergedFilesStatus();

    return (
      <div className={appCoreClassName}>
        <Sidebar
          isOpen={!!sidebarIsOpen}
          logFiles={processedLogFiles as ProcessedLogFiles}
          mergedFilesStatus={mergedFilesStatus}
          selectLogFile={this.selectLogFile}
          selectedLogFileName={selectedLogFileName}
        />
        <div id='content' className={logContentClassName}>
          <AppCoreHeader state={sleuthState} menuToggle={this.toggleSidebar} />
          {tableOrLoading}
        </div>
      </div>
    );
  }
}
