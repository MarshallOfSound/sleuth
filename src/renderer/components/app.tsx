import { resetState, sleuthState } from '../state/sleuth';
import { UserPreferences } from '../interfaces';
import { shouldIgnoreFile } from '../../utils/should-ignore-file';
import * as React from 'react';
import { ipcRenderer, remote } from 'electron';
import * as classNames from 'classnames';
import * as fs from 'fs-extra';
import * as path from 'path';

import { UnzippedFile, UnzippedFiles, Unzipper } from '../unzip';
import { Welcome } from './welcome';
import { CoreApplication } from './app-core';
import { MacTitlebar } from './mac-titlebar';
import { Preferences } from './preferences';
import { AppMenu } from '../menu';

const debug = require('debug')('sleuth:app');

export interface AppState {
  unzippedFiles: UnzippedFiles;
  userPreferences: UserPreferences;
}

export class App extends React.Component<undefined, Partial<AppState>> {
  public readonly menu: AppMenu = new AppMenu();

  constructor() {
    super();

    this.state = {
      unzippedFiles: []
    };

    localStorage.debug = 'sleuth*';

    const isDevMode = process.execPath.match(/[\\/]electron/);
    if (isDevMode) {
      try {
        (window as any).Perf = require('react-addons-perf');
      } catch (e) {
        debug(`Could not add React Perf`, e);
      }
    }

    this.openFile = this.openFile.bind(this);
  }

  /**
   * Should this component update?
   *
   * @param {undefined} _nextProps
   * @param {AppState} nextState
   */
  public shouldComponentUpdate(_nextProps: undefined, nextState: AppState) {
    const currentFiles = this.state.unzippedFiles || [];
    const nextFiles = nextState.unzippedFiles || [];

    if (currentFiles.length === 0 && nextFiles.length === 0) {
      return false;
    }

    return true;
  }

  /**
   * Alright, time to show the window!
   */
  public componentDidMount() {
    remote.getCurrentWindow().show();
    this.setupFileDrop();
  }

  /**
   * Whenever a file is dropped into the window, we'll try to open it
   */
  public setupFileDrop() {
    document.ondragover = document.ondrop = (event) => event.preventDefault();
    document.body.ondrop = (event) => {
      if (event.dataTransfer && event.dataTransfer.files.length > 0) {
        let url = event.dataTransfer.files[0].path;
        url = url.replace('file:///', '/');
        this.openFile(url);
      }

      event.preventDefault();
    };

    ipcRenderer.on('file-dropped', (_event: any, url: string) => this.openFile(url));
  }

  /**
   * Takes a rando string, quickly checks if it's a zip or not,
   * and either tries to open it as a file or as a folder. If
   * it's neither, we'll do nothing.
   *
   * @param {string} url
   * @returns {void}
   */
  public openFile(url: string): void {
    debug(`Received open-url for ${url}`);
    this.reset();

    const isZipFile = /[\s\S]*\.zip$/.test(url);
    if (isZipFile) {
      return this.openZip(url);
    }

    // Let's look at the url a little closer
    fs.stat(url).then((stats: fs.Stats) => {
      if (stats.isDirectory()) {
        return this.openDirectory(url);
      }
    });
  }

  /**
   * Takes a folder url as a string and opens it.
   *
   * @param {string} url
   */
  public openDirectory(url: string): void {
    debug(`Now opening directory ${url}`);
    this.reset();

    fs.readdir(url)
      .then((dir) => {
        const unzippedFiles: UnzippedFiles = [];
        const promises: Array<Promise<any>> = [];

        dir.forEach((fileName) => {
          if (shouldIgnoreFile(fileName)) return;

          const fullPath = path.join(url, fileName);
          debug(`Checking out file ${fileName}`);

          const promise = fs.stat(fullPath)
            .then((stats: fs.Stats) => {
              const file: UnzippedFile = { fileName, fullPath, size: stats.size };
              debug('Found file, adding to result.', file);
              unzippedFiles.push(file);
            });

          promises.push(promise);
        });

        Promise.all(promises).then(() => this.setState({ unzippedFiles }));
      });
  }

  /**
   * Takes a zip file url as a string and opens it.
   *
   * @param {string} url
   */
  public openZip(url: string): void {
    const unzipper = new Unzipper(url);
    unzipper.open()
      .then(() => unzipper.unzip())
      .then((unzippedFiles: UnzippedFiles) => this.setState({unzippedFiles}));
  }

  public reset() {
    this.setState({ unzippedFiles: [] });

    if (sleuthState.opened > 0) {
      resetState();
    }

    sleuthState.opened = sleuthState.opened + 1;
  }

  /**
   * Let's render this!
   *
   * @returns {JSX.Element}
   */
  public render(): JSX.Element {
    const { unzippedFiles } = this.state;
    const className = classNames('App', { Darwin: process.platform === 'darwin' });
    const titleBar = process.platform === 'darwin' ? <MacTitlebar /> : '';
    let content: JSX.Element | null = <Welcome openFile={this.openFile} />;

    if (unzippedFiles && unzippedFiles.length > 0) {
      content = <CoreApplication state={sleuthState} unzippedFiles={unzippedFiles} />;
    }

    return (
      <div className={className}>
        <Preferences state={sleuthState} />
        {titleBar}
        {content}
      </div>
    );
  }
}
