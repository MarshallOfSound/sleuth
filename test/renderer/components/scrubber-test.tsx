import * as React from 'react';
import { Scrubber } from '../../../src/renderer/components/scrubber';
import * as renderer from 'react-test-renderer';

it('scrubber renders correctly', () => {
  const tree = renderer.create(
    <Scrubber onResizeHandler={() => {}} elementSelector='window' />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});