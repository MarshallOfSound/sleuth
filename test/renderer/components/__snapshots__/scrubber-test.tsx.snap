// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrubber renders correctly 1`] = `
<div
  className="Scrubber"
  onMouseDown={[Function]}
/>
`;
